server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,152 @@
+        if (this.enabled) {
+            this.clusterService.addListener(clusterStateListener);
+            this.diskThresholdSettings.addListener(diskThresholdListener);
+        }
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ENABLED_SETTING, this::enable);
+    }
+
+    private void enable(boolean enabled) {
+        this.enabled = enabled;
+        if (this.enabled) {
+            clusterService.addListener(clusterStateListener);
+            diskThresholdSettings.addListener(diskThresholdListener);
+        } else {
+            clusterService.removeListener(clusterStateListener);
+            diskThresholdSettings.removeListener(diskThresholdListener);
+            publishedAfterElection = false;
+        }
+    }
+
+    private void updateHealthMetadataIfNecessary(ClusterChangedEvent event) {
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            if (event.localNodeMaster() && publishedAfterElection == false) {
+                submitHealthMetadata("health-metadata-update-master-election");
+            }
+            // If the node is not the elected master anymore
+            publishedAfterElection = event.localNodeMaster();
+        }
+    }
+
+    private void updateHealthMetadataIfNecessary() {
+        ClusterState clusterState = clusterService.state();
+        if (clusterState.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)
+            && clusterState.nodes().isLocalNodeElectedMaster()) {
+            submitHealthMetadata("health-metadata-update-threshold-change");
+        }
+    }
+
+    public static List<NamedXContentRegistry.Entry> getNamedXContentParsers() {
+        return List.of(
+            new NamedXContentRegistry.Entry(Metadata.Custom.class, new ParseField(HealthMetadata.TYPE), HealthMetadata::fromXContent)
+        );
+    }
+
+    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return List.of(
+            new NamedWriteableRegistry.Entry(Metadata.Custom.class, HealthMetadata.TYPE, HealthMetadata::new),
+            new NamedWriteableRegistry.Entry(NamedDiff.class, HealthMetadata.TYPE, HealthMetadata::readDiffFrom)
+        );
+    }
+
+    private HealthMetadata createHealthMetadata() {
+        return new HealthMetadata(HealthMetadata.DiskThresholds.createDiskThresholds(diskThresholdSettings));
+    }
+
+    private void submitHealthMetadata(String source) {
+        HealthMetadata localHealthMedata = createHealthMetadata();
+        var task = new UpdateHealthMetadataTask(localHealthMedata);
+        var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+        clusterService.submitStateUpdateTask(source, task, config, taskExecutor);
+    }
+
+    record UpdateHealthMetadataTask(HealthMetadata healthMetadata) implements ClusterStateTaskListener {
```

Comment:
I'm a bit confused by this record.

It's a task but doesn't have an execute method ? Should the `createHealthMetadata` be a sort of ... `upsertHealthMedata` and be executed inside the task's execute method?

Reply:
I see. What about this:
In the health metadata class we have classes that monitor the metadata settings. The task takes this monitors and in the execute method creates the current version of the HealthMetadata. 

I implemented this idea, in my latest commit. What do you think?

Reply:
Thanks for iterating on this.

Do we need this extra concept of "monitoring" though? Do we need the caching it provides for a particular use case?

Could the `HealthMetadataService` be the settings listener and trigger the necessary cluster state updates when they change? (apologies if I'm missing something)

Reply:
I do not think you are missing something. I was trying to optimize the case when we initialize the health metadata for the first time or when a master node gets elected. 

We could also parse the settings on the spot and have no cache, I was trying to avoid parsing them multiple times but on the other hand if this doesn't happen often maybe it's less important.

Do you think it would be cleaner to only track the settings?

Reply:
Settled offline, we expect that the changes will not be in high frequency so parsing the settings will not be a big burden and it remove the need of a cache.